examples/smbclient.py

```diff
@@ -120,16 +120,55 @@ def do_password(self, line):
         if self.loggedIn is False:
             logging.error("Not logged in")
             return
-        from getpass import getpass
-        newPassword = getpass("New Password:")
+
         rpctransport = transport.SMBTransport(self.smb.getRemoteHost(), filename = r'\samr', smb_connection = self.smb)
         dce = rpctransport.get_dce_rpc()
         dce.connect()                     
         dce.bind(samr.MSRPC_UUID_SAMR)
-        samr.hSamrUnicodeChangePasswordUser2(dce, '\x00', self.username, self.password, newPassword, self.lmhash, self.nthash)
-        self.password = newPassword
-        self.lmhash = None
-        self.nthash = None
+        
+        hashes = line.split(' ')[0]
+        if hashes != '':
+            lmhash, nthash = hashes.split(':')
+            # If a hash is received as a parameter  
+            try:
+                int(lmhash, 16)
+                int(nthash, 16)
+            except ValueError:
+                logging.error("Hashes are not valid")
+                return
+            
+            # Getting DomainID (SID)
+            serverHandle = samr.hSamrConnect(dce, self.smb.getServerName() + '\x00')['ServerHandle']
+            domainID = samr.hSamrLookupDomainInSamServer(dce, serverHandle, self.smb.getServerDomain())['DomainId']
+        
+            # Getting a DomainHandle
+            domainHandle = samr.hSamrOpenDomain(dce, serverHandle, domainId=domainID)['DomainHandle']   
+            
+            # Getting the User RID
+            userRID = samr.hSamrLookupNamesInDomain(dce, domainHandle, (self.username,))['RelativeIds']['Element'][0]
+                        
+            # Getting a UserHandle for the current user
+            userHandle = samr.hSamrOpenUser(dce, domainHandle, userId=userRID)['UserHandle']
+            
+            # Changing the password for the current user
+            if self.nthash != '':
```

Comment:
I'm confused about this line.. This would cover the case where hashes were sent, but no nthash was present. How can that happen?

Reply:
ohh.. whether the current credentials where provided as hash or password.. gotcha!... disregard.

Reply:
If you decided to change the current credentials (whether is was a plaintext or a hash)  for a new hash, then it would be:

True=> you authenticated using a hash, then call hSamrChangePasswordUser using that hash as oldPwdHashNT
False=> you authenticated using a plaintext pass, then call hSamrChangePasswordUser using the plaintext pass and let it calculate the hash by itself.

Does it make sense?

Reply:
:+1:

Reply:
hey @bransh 

What happens if the account is local to the target server?
I think you should handle that situation if this method also applies to local accounts.

By checking if the domain was specified through command line, you can call `hSamrLookupDomainInSamServer` with the `self.smb.getServerDomain()` of `self.smb.getServerName()` respectively. Haven't tested it tho.

Also.. I just ran it against a machine and it threw:
```
[-] SAMR SessionError: code: 0xc000006c - STATUS_PASSWORD_RESTRICTION - When trying to update a password, this status indicates that some password update rule has been violated. For example, the password may not meet length criteria.
```
Any idea? I'm wondering how/why the server threw that one..

Reply:
Regarding local accounts, you are right. I've just tried with self.smb.getServerName() in hSamrLookupDomainInSamServer and it gets the job done, though I will test it more later.

On the other hand, have you tried to change the password two times for that account? perhaps you have the minimum password age set to its default value of 1 (that is one condition we should care about).

Reply:
Gotcha.. will give it a try.. Now.. what about the `Enforce password history` policy? That one might prevent setting the password back.. most probably

Reply:
Just say you highlighting that in your Notes..